PhysicsTools/NanoAOD/plugins/LeptonInJetProducer.cc

```diff
@@ -0,0 +1,285 @@
+  std::tuple<float,float> calculateLSF(std::vector<fastjet::PseudoJet> iCParticles, std::vector<fastjet::PseudoJet> &ljets,
+				       float ilPt, float ilEta, float ilPhi, int ilId, double dr, int nsj) const;
+
+  edm::EDGetTokenT<pat::PackedCandidateCollection> srcPF_;
+  edm::EDGetTokenT<edm::View<pat::Jet>> srcJet_;
+  edm::EDGetTokenT<edm::View<pat::Electron>> srcEle_;
+  edm::EDGetTokenT<edm::View<pat::Muon>> srcMu_;
+
+};
+
+// ------------ method called to produce the data  ------------
+template <typename T>
+void
+LeptonInJetProducer<T>::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const
+{
+
+    // needs PFcandidate collection (srcPF), jet collection (srcJet), leptons collection 
+    edm::Handle<pat::PackedCandidateCollection> srcPF;
+    iEvent.getByToken(srcPF_, srcPF);
+    edm::Handle<edm::View<pat::Jet>> srcJet;
+    iEvent.getByToken(srcJet_, srcJet);
+    edm::Handle<edm::View<pat::Electron>> srcEle;
+    iEvent.getByToken(srcEle_, srcEle);   
+    edm::Handle<edm::View<pat::Muon>> srcMu;
+    iEvent.getByToken(srcMu_, srcMu);
+
+    unsigned int nJet = srcJet->size();
+    unsigned int nEle = srcEle->size();
+    unsigned int nMu  = srcMu->size();
+
+    std::vector<float> *vlsf3 = new std::vector<float>;
+    std::vector<float> *vlmd3 = new std::vector<float>;
+    std::vector<float> *vlep3pt = new std::vector<float>;
+    std::vector<float> *vlep3eta = new std::vector<float>;
+    std::vector<float> *vlep3phi = new std::vector<float>;
+    std::vector<int> *vlep3id = new std::vector<int>;
+
+    std::vector<float> *vlsf3match = new std::vector<float>;
+    std::vector<int> *vlep3idmatch = new std::vector<int>;
+
+    // Matching electron and pfcands  
+    const pat::PackedCandidate *ele_pfmatch = nullptr;
+    double pfDR = 0.05;
+    for (unsigned int il(0); il < nEle; il++) {
+      auto itLep = srcEle->ptrAt(il);
+      for (const pat::PackedCandidate &itPF : *srcPF) {
+	double dR = reco::deltaR(itLep->eta(), itLep->phi(), itPF.eta(), itPF.phi());
+	if(dR < pfDR) {
+	  pfDR = dR;
+          ele_pfmatch = &itPF;
+	}
+      }
+    } // Loop over pfcands
+    const pat::PackedCandidate *mu_pfmatch = nullptr;
+    pfDR = 0.05;
+    for (unsigned int il(0); il < nMu; il++) {
+      auto itLep = srcMu->ptrAt(il);
+      for (const pat::PackedCandidate &itPF : *srcPF) {
+        double dR = reco::deltaR(itLep->eta(), itLep->phi(), itPF.eta(), itPF.phi());
+        if(dR < pfDR) {
+          pfDR = dR;
+          mu_pfmatch = &itPF;
+        }
+      }
+    } // Loop over pfcands
+
+    // Find leptons in jets
+    for (unsigned int ij = 0; ij<nJet; ij++){
+      const pat::Jet &itJet = (*srcJet)[ij];
+      if(itJet.pt() <= 10) continue;
+      std::vector<fastjet::PseudoJet> lClusterParticles;
+      float lepPt(-1),lepEta(-1),lepPhi(-1);
+      //double dRconst(999);
+      int lepId(0);
+      for (unsigned int i0 = 0; i0 < itJet.numberOfSourceCandidatePtrs(); i0++) {
```

Comment:
is the navigation now working this way for fatjets or one should further navigate inside if the candidate is a subjet ? @rappoccio

Reply:
You should be able to change:
```
      for (unsigned int i0 = 0; i0 < itJet.numberOfSourceCandidatePtrs(); i0++) {
        const pat::PackedCandidate &itPF = dynamic_cast<const pat::PackedCandidate &>(*itJet.sourceCandidatePtr(i0));
	fastjet::PseudoJet pPart(itPF.px(),itPF.py(),itPF.pz(),itPF.energy());
```

to

```
      for (auto const & d : itJet->daughterPtrVector() ) {
          fastjet::PseudoJet p( d->px(), d->py(), d->pz(), d->energy() );
```

Reply:
@rappoccio I'm thinking I might need a dynamic_cast because I can't seem to access the properties of that candidate this way
` for (auto const & d : itJet.daughterPtrVector() ) {
	fastjet::PseudoJet p( d.px(), d.py(), d.pz(), d.energy() );`

I get the following errors:
` error: 'const class edm::Ptr<reco::Candidate>' has no member named 'px'
  fastjet::PseudoJet p( d.px(), d.py(), d.pz(), d.energy() );`

Am I missing something?

Reply:
`Ptr` is like a pointer, you have to do `->`, not `.`.